File: models/xgboost_classification_gradcam.py
"""
XGBoost with Train / Validation / External Validation
------------------------------
Author: Lucas Lacerda de Souza

Description:
    This script trains and evaluates an XGBoost model using all nuclear factors combined
    (multivariate model). It performs internal validation (train/validation split)
    and external validation using a separate dataset. The output includes a unified 
    performance plot comparing Training, Validation, and External Validation sets.

Dependencies:
    - xgboost
    - pandas
    - numpy
    - scikit-learn
    - matplotlib
    - seaborn
"""
import os
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.metrics import (
    accuracy_score, precision_score, recall_score, f1_score,
    roc_auc_score
)
from sklearn.model_selection import train_test_split
from xgboost import XGBClassifier


# =========================================================
# Load Datasets
# =========================================================
def load_data(train_path, test_path):
    train_df = pd.read_excel(train_path)
    test_df = pd.read_excel(test_path)

    features = [
        "Nucleus: Area",
        "Nucleus: Perimeter",
        "Nucleus: Circularity",
        "Nucleus: Eccentricity",
        "Nucleus: Hematoxylin OD mean"
    ]

    train_df = train_df[["Classe"] + features]
    test_df = test_df[["Classe"] + features]

    return train_df, test_df


# =========================================================
# XGBoost Training
# =========================================================
def train_xgboost(train_df, test_df, results_dir="Results_XGBoost_AllFactors", seed=123):
    os.makedirs(results_dir, exist_ok=True)

    X = train_df.drop(columns=["Classe"])
    y = train_df["Classe"]

    X_train, X_val, y_train, y_val = train_test_split(
        X, y, test_size=0.2, stratify=y, random_state=seed
    )

    X_test = test_df.drop(columns=["Classe"])
    y_test = test_df["Classe"]

    model = XGBClassifier(
        objective="binary:logistic",
        booster="gbtree",
        learning_rate=0.01,
        max_depth=8,
        subsample=0.9,
        colsample_bytree=0.9,
        n_estimators=3000,
        eval_metric="aucpr",
        random_state=seed,
        use_label_encoder=False
    )

    model.fit(
        X_train, y_train,
        eval_set=[(X_train, y_train), (X_val, y_val)],
        early_stopping_rounds=100,
        verbose=True
    )

    preds_train = model.predict_proba(X_train)[:, 1]
    preds_val = model.predict_proba(X_val)[:, 1]
    preds_test = model.predict_proba(X_test)[:, 1]

    return model, (X_train, y_train, preds_train), (X_val, y_val, preds_val), (X_test, y_test, preds_test)


# =========================================================
# Evaluation
# =========================================================
def compute_metrics(y_true, y_prob):
    y_pred = (y_prob > 0.5).astype(int)
    return {
        "Accuracy": accuracy_score(y_true, y_pred),
        "Precision": precision_score(y_true, y_pred, zero_division=0),
        "Recall": recall_score(y_true, y_pred, zero_division=0),
        "F1_Score": f1_score(y_true, y_pred, zero_division=0),
        "AUC": roc_auc_score(y_true, y_prob)
    }


# =========================================================
# Custom Metrics (Editable Section)
# =========================================================
def custom_metrics():
    return pd.DataFrame({
        "Set": ["Training", "Validation", "External Validation"],
        "Accuracy":  [0.926, 0.776, 0.817],
        "Precision": [0.937, 0.778, 0.834],
        "Recall":    [0.949, 0.767, 0.824],
        "F1_Score":  [0.926, 0.742, 0.819],
        "AUC":       [0.934, 0.737, 0.847]
    })


# =========================================================
# Visualization
# =========================================================
def plot_metrics(metrics_df, results_dir="Results_XGBoost_AllFactors"):
    metrics_long = metrics_df.melt(id_vars="Set", var_name="Metric", value_name="Value")
    metrics_long["Set"] = pd.Categorical(metrics_long["Set"],
                                         categories=["Training", "Validation", "External Validation"],
                                         ordered=True)

    sns.set(style="whitegrid", font_scale=1.2)
    plt.figure(figsize=(10, 6))
    ax = sns.barplot(data=metrics_long, x="Metric", y="Value", hue="Set",
                     palette=["#2166AC", "#67A9CF", "#D1E5F0"])
    for container in ax.containers:
        ax.bar_label(container, fmt="%.3f", label_type="edge", padding=3)
    plt.ylim(0, 1.05)
    plt.title("XGBoost Performance (All Nuclear Factors)\nTraining | Validation | External Validation",
              fontsize=15, weight="bold")
    plt.xlabel("Metric")
    plt.ylabel("Value")
    plt.legend(title="Dataset", loc="upper center", bbox_to_anchor=(0.5, 1.15), ncol=3)
    plt.tight_layout()

    save_path = os.path.join(results_dir, "all_datasets_metrics_customizable.png")
    plt.savefig(save_path, dpi=300)
    plt.close()
    print(f"Plot saved at: {save_path}")


# =========================================================
# Main Execution
# =========================================================
if __name__ == "__main__":
    train_path = "./data/XGBoost_Train.xlsx"
    test_path = "./data/XGBoost_Test.xlsx"
    results_dir = "./Results_XGBoost_AllFactors"

    train_df, test_df = load_data(train_path, test_path)

    model, train_res, val_res, test_res = train_xgboost(train_df, test_df, results_dir=results_dir)

    metrics_train = compute_metrics(train_res[1], train_res[2])
    metrics_val = compute_metrics(val_res[1], val_res[2])
    metrics_test = compute_metrics(test_res[1], test_res[2])

    metrics_df = pd.DataFrame([
        {"Set": "Training", **metrics_train},
        {"Set": "Validation", **metrics_val},
        {"Set": "External Validation", **metrics_test},
    ])

    # Optional: replace computed metrics with manually averaged ones
    # metrics_df = custom_metrics()

    print("\nFinal Metrics:")
    print(metrics_df.round(3))

    metrics_df.to_csv(os.path.join(results_dir, "xgboost_all_metrics.csv"), index=False)
    plot_metrics(metrics_df, results_dir=results_dir)

File: models/xgboost_classification_gradcam.R
# =========================================================
#  XGBoost Pipeline: Train / Validation / External Test
#  Multivariate Model Using All Nuclear Features
#  Unified Metric Plot (Train → Validation → External Test)
#  + Editable Section for Custom Metric Values
# =========================================================

# --- Load Required Libraries ---
packages <- c(
  "xgboost", "caret", "pROC", "MLmetrics", "dplyr",
  "ggplot2", "readxl", "scales", "reshape2"
)

for (pkg in packages) {
  if (!require(pkg, character.only = TRUE)) {
    install.packages(pkg, repos = "https://cran.r-project.org/")
    library(pkg, character.only = TRUE)
  }
}

# =========================================================
#  Load Datasets
# =========================================================
train_data <- read_excel("file.xlsx")
test_data  <- read_excel("file.xlsx")

features <- c(
  "Nucleus: Area",
  "Nucleus: Perimeter",
  "Nucleus: Circularity",
  "Nucleus: Eccentricity",
  "Nucleus: Hematoxylin OD mean"
)

train_data <- train_data[, c("Classe", features)]
test_data  <- test_data[,  c("Classe", features)]

# =========================================================
#  Split Dataset (80% Train / 20% Validation)
# =========================================================
set.seed(123)
train_index <- createDataPartition(train_data$Classe, p = 0.8, list = FALSE)

train_part <- train_data[train_index, ]
val_part   <- train_data[-train_index, ]

# --- Prepare Matrices ---
y_train <- train_part$Classe
X_train <- as.matrix(train_part[, -1])

y_val <- val_part$Classe
X_val <- as.matrix(val_part[, -1])

y_test <- test_data$Classe
X_test <- as.matrix(test_data[, -1])

dtrain <- xgb.DMatrix(data = X_train, label = y_train)
dval   <- xgb.DMatrix(data = X_val, label = y_val)
dtest  <- xgb.DMatrix(data = X_test, label = y_test)

watchlist <- list(train = dtrain, val = dval)

# =========================================================
#  Train XGBoost Model
# =========================================================
params <- list(
  objective = "binary:logistic",
  booster = "gbtree",
  eta = 0.01,
  max_depth = 8,
  subsample = 0.9,
  colsample_bytree = 0.9,
  eval_metric = "aucpr"
)

set.seed(123)
xgb_model <- xgb.train(
  params = params,
  data = dtrain,
  nrounds = 3000,
  watchlist = watchlist,
  early_stopping_rounds = 100,
  verbose = 1
)

# =========================================================
#  Evaluation: Train / Validation / External Test
# =========================================================
get_metrics <- function(y_true, y_prob) {
  pred_labels <- ifelse(y_prob > 0.5, 1, 0)
  conf_mat <- confusionMatrix(as.factor(pred_labels), as.factor(y_true), positive = "1")
  
  data.frame(
    Accuracy  = conf_mat$overall["Accuracy"],
    Precision = Precision(y_pred = pred_labels, y_true = y_true),
    Recall    = Recall(y_pred = pred_labels, y_true = y_true),
    F1_Score  = F1_Score(y_pred = pred_labels, y_true = y_true),
    AUC       = auc(roc(y_true, y_prob))
  )
}

train_pred <- predict(xgb_model, X_train)
val_pred   <- predict(xgb_model, X_val)
test_pred  <- predict(xgb_model, X_test)

train_metrics <- get_metrics(y_train, train_pred)
val_metrics   <- get_metrics(y_val, val_pred)
test_metrics  <- get_metrics(y_test, test_pred)

all_metrics <- rbind(
  cbind(Set = "Training", train_metrics),
  cbind(Set = "Validation", val_metrics),
  cbind(Set = "External Test", test_metrics)
)

# =========================================================
#  Visualization
#  Ordered: Training → Validation → External Test
# =========================================================
dir.create("Results_XGBoost_AllFeatures", showWarnings = FALSE)

all_long <- reshape2::melt(all_metrics, id.vars = "Set",
                           variable.name = "Metric", value.name = "Value")

all_long$Set <- factor(all_long$Set,
                       levels = c("Training", "Validation", "External Test"))

p <- ggplot(all_long, aes(x = Metric, y = Value, fill = Set)) +
  geom_bar(stat = "identity",
           position = position_dodge(width = 0.8),
           width = 0.6) +
  geom_text(aes(label = sprintf("%.3f", Value)),
            position = position_dodge(0.8),
            vjust = -0.5,
            size = 3.5) +
  scale_fill_manual(values = c("#2166AC", "#67A9CF", "#D1E5F0")) +
  ylim(0, 1.05) +
  labs(
    title = "XGBoost Performance (All Nuclear Features)",
    subtitle = "Training | Validation | External Test",
    x = "Metric",
    y = "Value",
    fill = "Dataset"
  ) +
  theme_minimal(base_size = 14) +
  theme(
    legend.position = "top",
    legend.title = element_text(face = "bold")
  )

print(p)

ggsave(
  "Results_XGBoost_AllFeatures/all_datasets_metrics.png",
  p, width = 9, height = 5.5, dpi = 300
)
